test/prototype/mx_formats/test_custom_cast.py

```diff
@@ -411,3 +422,168 @@ def test_fp6_e3m2_rounding(f32_val, f6_e3m2_enc, device):
+        + data_min
+    )
+
+    data_lp = f32_to_f4_e3m0_unpacked(data)
+    target_lp = torch.where((data >= 0) & (data <= F4_E3M0_MIN_NORMAL / 2), 0, 1).to(
+        torch.uint8
+    )
+    target_lp = torch.where(
+        data < -F4_E3M0_MIN_NORMAL / 2, 1 + 2**EBITS_F4_E3M0, data_lp
+    )
+
+    torch.testing.assert_close(
+        data_lp,
+        target_lp,
+        atol=0,
+        rtol=0,
+    )
+
+
+@pytest.mark.parametrize("hp_dtype", [torch.float32])
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+def test_float4_e2m1_underflow_use_stochastic_rounding(hp_dtype, device):
+    data_min = -F4_E3M0_MIN_NORMAL
+    data_max = F4_E3M0_MIN_NORMAL
+    data = (
+        torch.rand(1024, 1024, dtype=hp_dtype, device=device) * (data_max - data_min)
+        + data_min
+    )
+
+    data_lp = f32_to_f4_e3m0_unpacked(data, RoundingMode.STOCHASTIC)
+    target_lp = torch.where((data >= 0) & (data <= F4_E3M0_MIN_NORMAL / 2), 0, 1).to(
+        torch.uint8
+    )
+    target_lp = torch.where(
+        data < -F4_E3M0_MIN_NORMAL / 2, 1 + 2**EBITS_F4_E3M0, data_lp
+    )
+
+
+    torch.testing.assert_close(
+        data_lp,
+        target_lp,
+        atol=1,
+        rtol=0,
+    )
+
+    zeros_in_data_lp = (data_lp == 0).sum().item()
+    zeros_in_target_lp = (target_lp == 0).sum().item()
+
+    assert (
+        zeros_in_data_lp >= zeros_in_target_lp
+    ), f"stochastic rounding should have more non-zero values {zeros_in_data_lp} >= {zeros_in_target_lp}"
+
+
+@pytest.mark.parametrize("exp_range", list(range(-2, 4)))
+@pytest.mark.parametrize("hp_dtype", [torch.float32])
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+@pytest.mark.parametrize("sign", [1, -1])
+@pytest.mark.parametrize("rounding_mode", [False, True])
+def test_float4_e2m1_normal_cast(exp_range, hp_dtype, device, sign, rounding_mode):
+    if sign == 1:
+        data_min = pow(2, exp_range)
+        data_max = pow(2, exp_range + 1)
+    else:
+        data_min = -pow(2, exp_range + 1)
+        data_max = -pow(2, exp_range)
+
+    data = (
+        torch.rand(1024, 1024, dtype=hp_dtype, device=device) * (data_max - data_min)
+        + data_min
+    )
+
+    data_lp = f32_to_f4_e3m0_unpacked(data, rounding_mode).to(torch.float32)
+    if sign == 1:
+        data_lp = torch.pow(2, data_lp - F4_E3M0_EXP_BIAS)
+    else:
+        data_lp = -torch.pow(2, data_lp - F4_E3M0_EXP_BIAS - 8)
+
+    torch.testing.assert_close(
+        data_lp,
+        data,
+        atol=data_max - data_min,
+        rtol=0,
+    )
+
+
+@pytest.mark.parametrize("data_range", [1, 0.75, 0.5, 0.25, 0.125])
+@pytest.mark.parametrize("hp_dtype", [torch.float32, torch.bfloat16])
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+@pytest.mark.parametrize("block_size", [32])
+@pytest.mark.parametrize(
+    "rounding_mode", [RoundingMode.TIE_TO_EVEN, RoundingMode.STOCHASTIC]
+)
+def test_float4_e2m1_mx_qdq(data_range, hp_dtype, block_size, device, rounding_mode):
```

Comment:
this is testing `MXTensor`, can we move to `test/prototype/mx_formats/test_mx_tensor.py`?